### DISCUSSION TITLE
How to make fully self contained plugins?

### DISCUSSION THREAD

**Discussion by djowel** [free-audio/clap-wrapper#264] [Q&A]
Hi,

I followed the [Quick Start Guide](https://github.com/free-audio/clap-wrapper/wiki/Quick-Start-Guide-(Integrated)), and so far everything seems to work fine. However, what I have is currently VST3 land AU plugins that load. the resulting CLAP dynamically. What I want are fully self-contained plugins, as the quick guide hinted: 

"The VST3 loads the resulting CLAP dynamically, but there are other strategies to make fully self-contained plugins"

Can someone point me in the right direction or tell me how to do this?

Regards and thanks!

**Comment by baconpaul** [comment]
Sure

https://github.com/surge-synthesizer/conduit is a worked example of this. Basically you make your plugin into a static library then link it with each of the wrappers, including the clap, symmetrically.

**Reply by djowel** [reply]
Splendid! I'll have a look!

Awesome work, BTW!

**Reply by djowel** [reply]
So correct me if I am wrong. Basically:

1. Don't tagretize:  CLAP_WRAPPER_DONT_ADD_TARGETS
2. Do the VST3 and AU targets manually, statically linked to the clap entry points.

Something else I am missing?

**Reply by baconpaul** [reply]
That's basically it yeah. Use the cmake methods to put what you want into your vst3, and the cmake methods will pull in the wrapper. 

There's also a trick where the clap entry gets linked into every instance since it has the export symbol and in conduit at least we dlopen self to find the clap entry. There's some recent work on next to do that with a fully static clap entry.

**Reply by wadamek65** [reply]
Not sure if I understand this correctly, but is this possible if I already have a pre-built CLAP plugin written using Rust? What would I need to do in this case to statically link it to make the thick wrapper?

**Reply by baconpaul** [reply]
in that case you would want to not make a .clap with rust but instead make a static library which exports the three clap entry functions with well defined c names, then use the C++ linker to assemble with the shim of C++ to define the exported clap entry.

There's no requirement that the static clap library be written in c++; just that it resolves symbols the linker can find.

**Reply by wadamek65** [reply]
I see, thank you for the explanations @baconpaul ❤️ Is there perhaps an example of a similar integration or documentation on what steps exactly this would entail? I dived into the repository you linked previously and took a look at the CMake files but it's a little bit difficult for me to comprehend which is the thick-wrapper related part exactly. Apologies if the questions seem rather abstract as I can't ask better ones due to my unfamiliarity with C/C++.

**Reply by SG60** [reply]
@wadamek65 Did you manage to work this out? I'm also struggling to get a thick wrapper built from a Rust plugin.

**Reply by wadamek65** [reply]
@SG60 Unfortunately, no, I didn't even explore it. It wasn't a deal-breaker for me and a thin wrapper worked just fine. If you do manage to figure it out, I'd love to take a look though :)

**Reply by SG60** [reply]
@wadamek65 I've just uploaded an example repo that is working for me (using NIH-plug on the Rust side): https://github.com/SG60/rust-clap-wrapper-thick

**Comment by baconpaul** [comment]
So 

have rust not make a dyliv 
Have it make a static library which exports the three functions 
Then use the c+= linker and cmake to make the Dylib vst au etc like the clap first approach

I know how to do the last step if someone has a worked example of the first two 

but  y rust is weak enough that doing those would take me a bunch of time. And I don’t think doing it in Fortran or what not would help you all

**Reply by SG60** [reply]
So I think I've got the three functions exported from a Rust static library. However, I'm now getting loads of errors when building the CLAP wrapper.

The errors are like the following (but much longer):

```
Undefined symbols for architecture arm64:
  "_AudioComponentFindNext", referenced from:
      coreaudio::audio_unit::AudioUnit::new_with_flags::hb4c75c0f1d07c69a in libsg_plugin_1.a[450](coreaudio-094f061dbf5d319e.coreaudio.cf0130ad32841073-cgu.6.rcgu.o)
  "_AudioComponentInstanceDispose", referenced from:
      _$LT$coreaudio..audio_unit..AudioUnit$u20$as$u20$core..ops..drop..Drop$GT$::drop::hb1c178cf4e1014ac in libsg_plugin_1.a[450](coreaudio-094f061dbf5d319e.coreaudio.cf0130ad32841073-cgu.6.rcgu.o)
  "_AudioComponentInstanceNew", referenced from:
      coreaudio::audio_unit::AudioUnit::new_with_flags::hb4c75c0f1d07c69a in libsg_plugin_1.a[450](coreaudio-094f061dbf5d319e.coreaudio.cf0130ad32841073-cgu.6.rcgu.o)
  "_AudioObjectAddPropertyListener", referenced from:
      coreaudio::audio_unit::macos_helpers::RateListener::register::h4d3a9ea2e209f29f in libsg_plugin_1.a[445](coreaudio-094f061dbf5d319e.coreaudio.cf0130ad32841073-cgu.1.rcgu.o)
      coreaudio::audio_unit::macos_helpers::AliveListener::register::h1de08a045f382b46 in libsg_plugin_1.a[445](coreaudio-094f061dbf5d319e.coreaudio.cf0130ad32841073-cgu.1.rcgu.o)
  "_AudioObjectGetPropertyData", referenced from:
      coreaudio::audio_unit::macos_helpers::get_default_device_id::hf578fe9c37148039 in libsg_plugin_1.a[445](coreaudio-094f061dbf5d319e.coreaudio.cf0130ad32841073-cgu.1.rcgu.o)
      coreaudio::audio_unit::macos_helpers::get_audio_device_ids_for_scope::h2bf8335449ef5705 in libsg_plugin_1.a[445](coreaudio-094f061dbf5d319e.coreaudio.cf0130ad32841073-cgu.1.rcgu.o)
      coreaudio::audio_unit::macos_helpers::get_audio_device_supports_scope::h310b87ff4c4b0fd2 in libsg_plugin_1.a[445](coreaudio-094f061dbf5d319e.coreaudio.cf0130ad32841073-cgu.1.rcgu.o)
      coreaudio::audio_unit::macos_helpers::get_device_name::h39475bb8b8513ab2 in libsg_plugin_1.a[445](coreaudio-094f061dbf5d319e.coreaudio.cf0130ad32841073-cgu.1.rcgu.o)
      coreaudio::audio_unit::macos_helpers::get_device_name::h39475bb8b8513ab2 in libsg_plugin_1.a[445](coreaudio-094f061dbf5d319e.coreaudio.cf0130ad32841073-cgu.1.rcgu.o)
      coreaudio::audio_unit::macos_helpers::set_device_sample_rate::hf2bdb2dce3d814ef in libsg_plugin_1.a[445](coreaudio-094f061dbf5d319e.coreaudio.cf0130ad32841073-cgu.1.rcgu.o)
      coreaudio::audio_unit::macos_helpers::set_device_sample_rate::hf2bdb2dce3d814ef in libsg_plugin_1.a[445](coreaudio-094f061dbf5d319e.coreaudio.cf0130ad32841073-cgu.1.rcgu.o)
      ...
```

The C++ file that I am using for the entrypoint is as follows. The `init`, `deinit`, and `get_factory` functions are exported from the Rust staticlib.

```cpp
/*
 * clap entry
 *
 * This uses extern versions of the entry methods from a static library to create
 * a working exported C linkage clap entry point
 */

#include <cstring>

#include <clap/clap.h>

#include "GENERATED/sg-plugin-1-bindings.h"

extern "C"
{
#ifdef __GNUC__
#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wattributes"  // other peoples errors are outside my scope
#endif

const CLAP_EXPORT struct clap_plugin_entry clap_entry = {CLAP_VERSION, init, deinit, get_factory};

#ifdef __GNUC__
#pragma GCC diagnostic pop
#endif
}
```

Do you have any idea what the issue here might be?

**Reply by SG60** [reply]
I can try and put together an example repo where it isn't working if you need.

**Reply by SG60** [reply]
Got it working now! A silly mistake – I needed to specify the dynamic libraries manually that the Rust static library depends on when linking with CMake, using something like this:

```cmake
target_link_libraries(${CLAP_TARGET} PUBLIC
            "-framework Foundation"
            "-framework CoreFoundation"
            "-framework AppKit"
            "-framework AudioToolbox"
            "-framework Quartz"
            "-framework CoreAudio"
            "-framework AudioUnit"
            "-framework CoreMIDI")
```

Thanks for the help!

**Reply by baconpaul** [reply]
Yeah I was about to say “missing frameworks at link stage?” - awesome

if you have a small example rust repo and would be happy to do so we could add it as an example in clap wrapper or as a standalone repo on free audio to help folks or point to it in our doc

glad it worked! I knew it should theoretically but nice to see it in reality

**Reply by SG60** [reply]
I've just uploaded a small example project here: https://github.com/SG60/rust-clap-wrapper-thick

It should build according to the instructions in the readme if you have Rust and Cbindgen installed.

**Reply by wadamek65** [reply]
@SG60 Awesome work! Thank you for sharing this, I'll definitely be taking a look 👀

**Reply by CrushedPixel** [reply]
Thanks a ton for sharing this, @SG60, this is a great starting point :)

**Reply by SG60** [reply]
I haven't tested this on Windows or Linux, so it currently only works on MacOS. [The list of frameworks in CMakeLists.txt needed by the Rust staticlib at the link stage](https://github.com/SG60/rust-clap-wrapper-thick/blob/00553d146566807a47e5b54b07e9edf4af83cb91/clap-wrapper-cmake/CMakeLists.txt#L70-L79) will need adjusting for each platform.

**Reply by CrushedPixel** [reply]
I set up a minimal example using clack (safe Rust CLAP bindings) that follows this approach: https://github.com/CrushedPixel/rust-clap-first-example

This also shows how to expose multiple plugins from the same binary, for AUv2 you need to use the clap-wrapper's AUv2 Factory Extension for that!

This doesn't use cbindgen, @SG60 it's not really needed since you only export the CLAP entrypoints under a constant name.

**Comment by djowel** [comment]
Thank you all for your inputs!  Much appreciated!
